File: regxmllib/src/com/sandflow/smpte/util/UUID.java
/*
 * Copyright (c) 2014, Pierre-Anthony Lemieux (pal@sandflow.com)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.sandflow.smpte.util;

import java.util.Arrays;
import java.util.regex.Pattern;

/**
 *
 * @author Pierre-Anthony Lemieux (pal@sandflow.com)
 */
public class UUID {

    private byte[] value;

    public UUID(byte[] uuid) {
        this.value = java.util.Arrays.copyOf(uuid, 16);
    }

    public byte[] getValue() {
        return this.value;
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 37 * hash + Arrays.hashCode(this.value);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final UUID other = (UUID) obj;
        if (!Arrays.equals(this.value, other.value)) {
            return false;
        }
        return true;
    }

    private final static Pattern URN_PATTERN = Pattern.compile("urn:uuid:[a-fA-F0-9]{8}-(?:[a-fA-F0-9]{4}-){3}[a-fA-F0-9]{12}");

    public static UUID fromURN(String val) {

        /* TODO: should this throw an exception */
        byte[] uuid = new byte[16];

        if (URN_PATTERN.matcher(val).matches()) {

            
        int inoff = 0;
        int outoff = 9;

        for (int j = 0; j < 4; j++) {

            uuid[inoff++] = (byte) Integer.parseUnsignedInt(val.substring(outoff, outoff + 2), 16);
            
            outoff += 2;
            
        }

        for (int i = 0; i < 3; i++) {

            outoff++;

            for (int j = 0; j < 2; j++) {

                uuid[inoff++] = (byte) Integer.parseUnsignedInt(val.substring(outoff, outoff + 2), 16);
                outoff += 2;
            }
        }

        outoff++;

        for (int j = 0; j < 6; j++) {

                uuid[inoff++] = (byte) Integer.parseUnsignedInt(val.substring(outoff, outoff + 2), 16);
                outoff += 2;
        }
            
            

            return new UUID(uuid);

        } else {

            return null;

        }

    }

    final static char[] HEXMAP = "0123456789abcdef".toCharArray();
    final static char[] URNTEMPLATE = "urn:uuid:3e0993c0-66e0-11e4-9803-0800200c9a66".toCharArray();

    @Override
    public String toString() {

        char[] out = Arrays.copyOf(URNTEMPLATE, URNTEMPLATE.length);

        int inoff = 0;
        int outoff = 9;

        for (int j = 0; j < 4; j++) {

            int v = value[inoff++] & 0xFF;
            out[outoff++] = HEXMAP[v >>> 4];
            out[outoff++] = HEXMAP[v & 0x0F];

        }

        for (int i = 0; i < 3; i++) {

            outoff++;

            for (int j = 0; j < 2; j++) {

                int v = value[inoff++] & 0xFF;
                out[outoff++] = HEXMAP[v >>> 4];
                out[outoff++] = HEXMAP[v & 0x0F];

            }
        }

        outoff++;

        for (int j = 0; j < 6; j++) {

            int v = value[inoff++] & 0xFF;
            out[outoff++] = HEXMAP[v >>> 4];
            out[outoff++] = HEXMAP[v & 0x0F];

        }

        return new String(out);
    }

}

File: regxmllib/src/main/cpp/com/sandflow/smpte/util/UUID.cpp
/*
* Copyright (c), Pierre-Anthony Lemieux (pal@palemieux.com)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
*
* * Redistributions of source code must retain the above copyright notice, this
*   list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright notice,
*   this list of conditions and the following disclaimer in the documentation
*   and/or other materials provided with the distribution.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*/

#include "UUID.h"
#include <sstream>  
#include <iomanip>
#include <cstdlib>
#include <cstdio>
#include <stdexcept>

namespace rxml {

	static const std::string UUID_URN_PREFIX = "urn:uuid:";

	bool lowerCmp(char i, char j) {
		return (::towlower(i) == ::towlower(j));
	}

	bool UUID::urnToBytes(const std::string &urn, unsigned char(&uuid)[16]) {

		if (urn.size() != 45) return false;

		if (std::equal(urn.begin(),
			urn.begin() + UUID_URN_PREFIX.size(),
			UUID_URN_PREFIX.begin(), lowerCmp) != true) return false;

		/* process as UUID (urn:uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6) */

		int pos = 9;

		for (int i = 0; i < 16; i++) {

			uuid[i] = std::strtol(urn.substr(pos, 2).c_str(), NULL, 16) & 0xff;

			if (i == 3 || i == 5 || i == 7 || i == 9) {
				pos += 3;
			} else {
				pos += 2;
			}

		}

		return true;

	}


	std::string UUID::bytesToString(const unsigned char uuid[16]) {

		char str[128];

		int r = snprintf(str, sizeof(str), "urn:uuid:%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
			uuid[0], uuid[1], uuid[2], uuid[3],
			uuid[4], uuid[5], uuid[6], uuid[7],
			uuid[8], uuid[9], uuid[10], uuid[11],
			uuid[12], uuid[13], uuid[14], uuid[15]
		);

		if (r < 0) throw std::runtime_error("UUID::bytesToString failed.");

		return str;

	}
}